File: client/scala/libs/spark.ml/src/main/scala-2.11/edu/mit/csail/db/ml/modeldb/client/event/ProjectEvent.scala
package edu.mit.csail.db.ml.modeldb.client.event

import com.twitter.util.Await
import edu.mit.csail.db.ml.modeldb.client.ModelDbSyncer
import modeldb.ModelDBService.FutureIface

/**
  * Stores a new project on the server.
  * @param project - The project to store.
  */
case class ProjectEvent(project: modeldb.Project) extends ModelDbEvent {
  /**
    * Store the project on the server.
    * @param client - The client that exposes the functions that we
    *               call to store objects in the ModelDB.
    * @param mdbs - The ModelDbSyncer, included so we can update the ID
    *             mappings after syncing.
    */
  override def sync(client: FutureIface, mdbs: Option[ModelDbSyncer]): Unit = {
    val res = Await.result(client.storeProjectEvent(modeldb.ProjectEvent(project)))
    mdbs.get.project = mdbs.get.project.copy(id = res.projectId)
  }
}

File: client/python/modeldb/events/ProjectEvent.py
"""
Store Project on server.
"""
from modeldb.events.Event import Event
from ..thrift.modeldb import ttypes as modeldb_types


class ProjectEvent(Event):
    """
    Class for creating and storing ProjectEvents
    """

    def __init__(self, project):
        self.project = project

    def make_event(self, syncer):
        """
        Constructs a thrift ProjectEvent object with appropriate fields.
        """
        return modeldb_types.ProjectEvent(self.project)

    def sync(self, syncer):
        """
        Stores ProjectEvent on the server.
        """
        thrift_client = syncer.client
        res = thrift_client.storeProjectEvent(self.make_event(syncer))
        syncer.project.id = res.projectId
